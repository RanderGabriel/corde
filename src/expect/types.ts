import {
  EmojiLike,
  MessageIdentifier,
  MessageEditedIdentifier,
  MessageEmbedLike,
  ChannelLocation,
  RoleIdentifier,
  TestFunctionType,
  CordeBotLike,
} from "../types";
import { ColorResolvable } from "discord.js";
import { Colors, RolePermission } from "../utils";

export interface ExpectTestBaseParams {
  cordeBot: CordeBotLike;
  command: string | number | boolean;
  isNot: boolean;
  timeout: number;
  isCascade: boolean;
  channelId?: string;
  guildId?: string;
}
export interface ExpectTestParams extends ExpectTestBaseParams {
  testName: string;
}

export type MayReturnMatch = Matches<any> | void;

/**
 * Defines all functions that can be used
 * to check a bot reaction of a command.
 *
 */
export interface MessageMatches<TReturn extends MayReturnMatch> {
  /**
   * Defines the message expected to be returned by a
   * command.
   *
   * @param expect A message returned by a bot after invoke a command
   * @since 1.0
   */
  toReturn(expect: bigint | boolean | number | string): TReturn;
  /**
   * Defines the message expected to be returned by a
   * command.
   *
   * @param expect A message returned by a bot after invoke a command
   * @since 1.0
   */
  toReturn(expect: MessageEmbedLike): TReturn;

  /**
   * Defines the message expected to be returned by a
   * command in a specific channel and guild(optional).
   *
   * @param expect A message returned by a bot after invoke a command
   * @param channelId Id of the channel where the message must be sent to.
   * @param guildId Id of the guild that contains the channel where the message must be sent to.
   * @since 3.0
   */
  toReturnInChannel(
    expect: bigint | boolean | number | string | MessageEmbedLike,
    channelId: string,
    guildId?: string,
  ): TReturn;

  /**
   * Defines the message expected to be returned by a
   * command in a specific channel and guild(optional).
   *
   * @param expect A message returned by a bot after invoke a command.
   * @param channelLocation Object with to holds the id of the channel and of the guild (optional).
   * @since 3.0
   */
  toReturnInChannel(
    expect: bigint | boolean | number | string | MessageEmbedLike,
    channelLocation: ChannelLocation,
  ): TReturn;

  /**
   * Defines [reactions](https://discordjs.guide/popular-topics/reactions.html#reacting-to-messages)
   * that must be added to the command message.
   *
   * @param emojis Single or list of reactions that must be added to a message.
   * It can be **emojis** or [custom emojis](https://support.discord.com/hc/en-us/articles/360036479811-Custom-Emojis).
   *
   * @example
   *
   * // Checks if the first message sent by the testing bot receives emojis in reactions.
   * expect("emoji").toAddReaction(['😄']);
   * expect("emoji").toAddReaction([{ name: '😄' }]);
   * expect("emoji").toAddReaction([{ name: '😄' }]);
   * expect("emoji").toAddReaction(['🍊', { name: '😄' }]);
   * expect("emoji").toAddReaction(['🍊', { name: '😄' }]);
   *
   * // Checks if a specific message receives emojis in reactions.
   * expect("emoji").toAddReaction(['😄'], { id: '96008815106887111' });
   * expect("emoji").toAddReaction([{ name: '😄' }], { id: '96008815106887111' });
   *
   * // This example will find for a message with content: 'message text'
   * expect("emoji").toAddReaction([{ name: '😄' }], { name: 'message text' });
   * expect("emoji").toAddReaction(['🍊', { name: '😄' }], { id: 'message text' });
   * expect("emoji").toAddReaction(['🍊', { name: '😄' }], { id: 'message text' });
   *
   * // Its also possible to search the message directly by it's id:
   * expect("emoji").toAddReaction(['😄'], '96008815106887111');
   *
   * @since 1.0
   */
  toAddReaction(emojis: string[]): TReturn;
  toAddReaction(emojis: EmojiLike[]): TReturn;
  toAddReaction(emojis: (string | EmojiLike)[]): TReturn;
  toAddReaction(emojis: EmojiLike[], messageIdentifier: string): TReturn;
  toAddReaction(emojis: EmojiLike[], messageIdentifier: MessageIdentifier): TReturn;
  toAddReaction(emojis: (string | EmojiLike)[], messageIdentifier: string): TReturn;
  toAddReaction(emojis: (string | EmojiLike)[], messageIdentifier: MessageIdentifier): TReturn;
  toAddReaction(emojis: string[], messageIdentifier: string): TReturn;
  toAddReaction(emojis: string[], messageIdentifier: MessageIdentifier): TReturn;

  /**
   * Check if a command removes a list of reactions from the last message sent
   * by someone who is not the testing bot or corde's bot.
   *
   * @param emojis Reactions to check if were removed. And/or reactions like (reactions that are customized).
   * @example
   *
   * toRemoveReaction([{ id: "12321" }, "😆"]);
   * toRemoveReaction([{ id: "12321" }, "😆"], "31241212512");
   *
   * @since 2.0
   */
  toRemoveReaction(emojis: EmojiLike[], messageId?: string): TReturn;
  /**
   * Check if a command removes a list of reactions from the last message sent
   * by someone who is not the the testing bot or corde's bot.
   *
   * @param emojis Reactions to check if were removed. And/or reactions like (reactions that are customized).
   * @param messageIdentifier Object with **id** or **content** of the message.
   * @example
   *
   * toRemoveReaction([{ id: "12321" }, { name: "😆" }]);
   * toRemoveReaction([{ id: "12321" }, { name: "😆" }], { id: "1232142"});
   *
   * @since 2.0
   */
  toRemoveReaction(emojis: EmojiLike[], messageIdentifier?: MessageIdentifier): TReturn;
  /**
   * Check if a command removes a list of reactions from the last message sent
   * by someone who is not the the testing bot or corde's bot.
   *
   * @param emojis Reactions to check if were removed. And/or reactions like (reactions that are customized).
   * @param messageId Id of the message.
   * @example
   *
   * toRemoveReaction([{ id: "12321" }, "😆"]);
   * toRemoveReaction([{ id: "12321" }, "😆"], "4121512121");
   *
   * @since 2.0
   */
  toRemoveReaction(emojis: (string | EmojiLike)[], messageId?: string): TReturn;
  /**
   * Check if a command removes a list of reactions from the last message sent
   * by someone who is not the the testing bot or corde's bot.
   *
   * @param emojis Reactions to check if were removed. And/or reactions like (reactions that are customized).
   * @param messageIdentifier Object with **id** or **content** of the message.
   * @example
   *
   * toRemoveReaction([{ id: "12321" }, "😆"]);
   * toRemoveReaction([{ id: "12321" }, "😆"], { id: "4121512121" });
   *
   * @since 2.0
   */
  toRemoveReaction(emojis: (string | EmojiLike)[], messageIdentifier?: MessageIdentifier): TReturn;
  /**
   * Check if a command removes a list of reactions from the last message sent
   * by someone who is not the the testing bot or corde's bot.
   *
   * @param emojis Reactions to check if were removed
   * @param messageId Id of the message.
   * @example
   *
   * toRemoveReaction(["😆"]);
   * toRemoveReaction(["😆"], { id: "13124124" });
   *
   * @since 2.0
   */
  toRemoveReaction(emojis: string[], messageId?: string): TReturn;
  /**
   * Check if a command removes a list of reactions from the last message sent
   * by someone who is not the the testing bot or corde's bot.
   *
   * @param emojis Reactions to check if were removed.
   * @param messageIdentifier Object with **id** or **content** of the message.
   * @example
   *
   * toRemoveReaction(["😆", "🥱"]);
   * toRemoveReaction(["😆", "🥱"], { id: "1234871972" });
   *
   * @since 2.0
   */
  toRemoveReaction(emojis: string[], messageIdentifier?: MessageIdentifier): TReturn;

  /**
   * Verify if a command pinned a message.
   *
   * @param messageId Id of the message
   * @since 2.0
   */
  toPin(messageId: string): TReturn;
  /**
   * Verify if a command pinned a message.
   *
   * @param messageIdentifier Object with **id** or **content** of the message.
   * @since 2.0
   */
  toPin(messageIdentifier: MessageIdentifier): TReturn;

  /**
   * Verify if a command unpinned a message.
   *
   * @param messageId Id of the message
   * @since 2.0
   */
  toUnPin(messageId: string): TReturn;
  /**
   * Verify if a command unpinned a message.
   *
   * @param message Object with **id** or **content** of the message.
   * @since 2.0
   */
  toUnPin(messageIdentifier: MessageIdentifier): TReturn;

  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **string** value for the message.
   * @param messageIdentifier Data object with the **id** or **oldContent** of the message.
   * @since 1.0
   */
  toEditMessage(newValue: string, messageIdentifier?: MessageEditedIdentifier): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **string** value for the message.
   * @param messageId Id of the message.
   * @since 1.0
   */
  toEditMessage(newValue: string, messageId?: string): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **number** value for the message.
   * @param messageIdentifier Data object with the **id** or **oldContent** of the message.
   * @since 1.0
   */
  toEditMessage(newValue: number, messageIdentifier?: MessageEditedIdentifier): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **number** value for the message.
   * @param messageId Id of the message.
   * @since 1.0
   */
  toEditMessage(newValue: number, messageId?: string): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **bigint** value for the message.
   * @param messageIdentifier Data object with the **id** or **oldContent** of the message.
   * @since 1.0
   */
  toEditMessage(newValue: bigint, messageIdentifier?: MessageEditedIdentifier): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **bigint** value for the message.
   * @param messageId Id of the message.
   * @since 1.0
   */
  toEditMessage(newValue: bigint, messageId?: string): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **boolean** value for the message.
   * @param messageId Id of the message.
   * @since 1.0
   */
  toEditMessage(newValue: boolean, messageId?: string): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **boolean** value for the message.
   * @param messageIdentifier Data object with the **id** or **oldContent** of the message.
   * @since 1.0
   */
  toEditMessage(newValue: boolean, messageIdentifier?: MessageEditedIdentifier): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **MessageEmbedLike** value for the message.
   * @param messageIdentifier Data object with the **id** or **oldContent** of the message.
   * @since 1.0
   */
  toEditMessage(newValue: MessageEmbedLike, messageIdentifier?: MessageEditedIdentifier): TReturn;
  /**
   * Verify if a command edited a message.
   *
   * @param newValue New **MessageEmbedLike** value for the message.
   * @param messageId Data object with the **id** or **oldContent** of the message.
   * @since 1.0
   */
  toEditMessage(newValue: MessageEmbedLike, messageId?: string): TReturn;
}

/**
 * Tests for a **Role** structure.
 */
export interface RoleMatches<TReturn extends MayReturnMatch> {
  /**
   * Check if a command changed a role color.
   *
   * @param color The new color for the role.
   * @param roleId Id of the role.
   * @since 2.0
   */
  toSetRoleColor(color: ColorResolvable, roleId: string): TReturn;
  /**
   * Check if a command changed a role color.
   *
   * @param color The new color for the role.
   * @param roleIdentifier Object with the **id** or the **name** of the role.
   * @since 2.0
   */
  toSetRoleColor(color: ColorResolvable, roleIdentifier: RoleIdentifier): TReturn;
  /**
   * Check if a command changed a role color.
   *
   * @param color Color enum in hexadecimal format.
   * @param roleId Object with the **id** or the **name** of the role.
   * @since 2.0
   */
  toSetRoleColor(color: Colors, roleId: string): TReturn;
  /**
   * Check if a command changed a role color.
   *
   * @param color Color enum in hexadecimal format.
   * @param roleId Object with the **id** or the **name** of the role.
   * @since 2.0
   */
  toSetRoleColor(color: Colors, roleIdentifier: RoleIdentifier): TReturn;

  /**
   * Check if a command deletes a role.
   *
   * @param roleId Id of the role.
   * @since 2.0
   */
  toDeleteRole(roleId: string): TReturn;

  /**
   * Check if a command deletes a role.
   *
   * @param roleIdentifier Object with **id** or **name** of the role.
   * @since 2.0
   */
  toDeleteRole(roleIdentifier: RoleIdentifier): TReturn;

  /**
   * Check if a command defines a role as mentionable.
   *
   * @param mentionable If the role can or can not be mentionable.
   * @param roleId Id of the role.
   * @since 2.0
   */
  toSetRoleMentionable(mentionable: boolean, roleId: string): TReturn;
  /**
   * Check if a command define a role as mentionable.
   *
   * @param mentionable If the role can or can not be mentionable.
   * @param roleIdentifier Object with **id** or **name** of the role.
   * @since 2.0
   */
  toSetRoleMentionable(mentionable: boolean, roleIdentifier: RoleIdentifier): TReturn;

  /**
   * Check if a commend defines a role as a hoist.
   *
   * @param hoist if this role is pinned in the user listing.
   * @param roleId Id of the role.
   *
   * @description Discord provides two methods of displaying roles; hoisted and standard.
   * The role hierarchy is visibly clear to server members; roles are sorted and displayed
   * based on which role is higher in the role management menu.
   *
   * However, in a standard configuration, users are sorted alphabetically, meaning someone
   * with the highest role will be sorted wherever their name exists in the alphabet.
   *
   * Source from [discord support](https://support.discord.com/hc/en-us/community/posts/360060076751-Un-hoisted-Role-Hierarchy).
   * @see https://discord.com/developers/docs/topics/permissions
   * @since 2.0
   */
  toSetRoleHoist(hoist: boolean, roleId: string): TReturn;
  /**
   * Check if a commend define a role as hoist.
   *
   * @param hoist if this role is pinned in the user listing.
   * @param roleIdentifier Object with **id** or **name** of the role.
   *
   * @description Discord provides two methods of displaying roles; hoisted and standard.
   * The role hierarchy is visibly clear to server members; roles are sorted and displayed
   * based on which role is higher in the role management menu.
   *
   * However, in a standard configuration, users are sorted alphabetically, meaning someone
   * with the highest role will be sorted wherever their name exists in the alphabet.
   *
   * Source from [discord support](https://support.discord.com/hc/en-us/community/posts/360060076751-Un-hoisted-Role-Hierarchy).
   * @see https://discord.com/developers/docs/topics/permissions
   * @since 2.0
   */
  toSetRoleHoist(hoist: boolean, roleIdentifier: RoleIdentifier): TReturn;

  /**
   * Check if a command renames a role.
   *
   * @param newName new name of the role
   * @param roleId Id of the role.
   * @since 2.0
   */
  toRenameRole(newName: string, roleId: string): TReturn;
  /**
   * Check if a command rename a role.
   *
   * @param newName new name of the role
   * @param roleIdentifier Object with **id** or **name** of the role.
   * @since 2.0
   */
  toRenameRole(newName: string, roleIdentifier: RoleIdentifier): TReturn;

  /**
   * Check if a command changes a role's position.
   *
   * @param newPosition The new position of the role.
   * @param roleId Id of the role.
   *
   * @description Role's maximum value depends on the number of roles the guild Have.
   * So, if there are only 3 roles (including the default *everyone*), the maximum
   * position that a role can have is 2 (The count begins with 0, So: 0, 1, 2).
   *
   * @see https://discord.com/developers/docs/topics/permissions
   * @since 2.0
   */
  toSetRolePosition(newPosition: number, roleId: string): TReturn;
  /**
   * Check if a command change a role's position.
   *
   * @param newPosition The new position of the role.
   * @param roleIdentifier Object with **id** or **name** of the role.
   *
   * @description Role's maximum value depends on the number of roles the guild has.
   * So, if there are only 3 roles (including the default *everyone*), the maximum
   * position that a role can have is 2 (The count begins with 0, So: 0, 1, 2).
   *
   * @see https://discord.com/developers/docs/topics/permissions
   * @since 2.0
   */
  toSetRolePosition(newPosition: number, roleIdentifier: RoleIdentifier): TReturn;

  /**
   * Check if a command change the
   * [permissions](https://discord.com/developers/docs/topics/permissions#permissions-bitwise-permission-flags)
   * of a role.
   *
   * @param roleId Id of the role.
   * @param permissions List of permissions allowed by Discord.
   *
   * @see https://discord.com/developers/docs/topics/permissions#permissions
   * @since 2.0
   */
  toSetRolePermission(roleId: string, ...permissions: RolePermission[]): TReturn;
  /**
   * Check if a command change the
   * [permissions](https://discord.com/developers/docs/topics/permissions#permissions-bitwise-permission-flags)
   * of a role.
   *
   * @param roleIdentifier Object with **id** or **name** of the role.
   * @param permissions List of permissions allowed by Discord.
   *
   * @see https://discord.com/developers/docs/topics/permissions#permissions
   * @since 2.0
   */
  toSetRolePermission(roleIdentifier: RoleIdentifier, ...permissions: RolePermission[]): TReturn;
}

export interface MacherContructorArgs {
  commandName: unknown;
  isNot?: boolean;
  channelId?: string;
  guildId?: string;
  isCascade?: boolean;
}

export interface CascadeMatch {
  todoInCascade(...tests: TestFunctionType[]): void;
}

/**
 * Defines all functions that can be used
 * to check a bot reaction of a command.
 */
export type Matches<TReturn extends MayReturnMatch> = MessageMatches<TReturn> &
  RoleMatches<TReturn>;

/**
 * Defines the initial value of expectations from
 * **command** function. It includes all matches and
 * the *not* statement. Witch will deny the executed match
 */
export interface IsNot<TMatchesResponse extends any> {
  /**
   * Defines that a command should **not** do something.
   * Use this if you can not precise what response a command will throw,
   * But know what it **can not** throw.
   */
  not: TMatchesResponse;
}

export interface SetChannelMatchers<TReturn extends MayReturnMatch> {
  /**
   * Specify a channel where tests will be **validated** on.
   *
   * @example
   *
   * // Suposing that the config is:
   * const cordeConfig = {
   *  channelId: "123",
   *  botPrefix: "!"
   * };
   *
   * expect("ping").inChannel("321").toReturn("pong");
   *
   * @description This will send the message "!ping" in the channel defined in configs("123"),
   * and check if a message with content "pong" will be sent to the text channel of if "321".
   *
   * @param id Id of the channel
   */
  inChannel(id: string): IsNot<MessageMatches<TReturn>> & MessageMatches<TReturn>;
}

export interface SetGuildMatchers<TReturn extends MayReturnMatch> {
  /**
   * Specify a guild where tests will be **validated** in.
   *
   * @example
   *
   * // Suposing that the config is:
   * const cordeConfig = {
   *  guildId: "123",
   *  botPrefix: "!"
   * };
   *
   * expect("ping").inGuild("321").toReturn("pong");
   *
   * @description This will send the message "!ping" in the channel defined in configs("123"),
   * and check if a message with content "pong" will be sent to the text channel of if "321".
   *
   * @param id Id of the channel
   */
  inGuild(id: string): IsNot<RoleMatches<TReturn>> & RoleMatches<TReturn>;
}

export type AllMatches<TReturn extends MayReturnMatch> = SetChannelMatchers<TReturn> &
  SetGuildMatchers<TReturn> &
  IsNot<Matches<TReturn>> &
  Matches<TReturn>;

export interface Expect extends AllMatches<any> {
  /**
   * Receives which command will be tested.
   *
   * Do not inform the command prefix if
   * it's already informed in **configs**
   *
   * @param commandNameResolvable Command name. (Empty strings will resolve failed test)
   *
   * @returns An object with all possible tests to be done
   * in the bot.
   *
   * @since 1.0
   */
  <T extends (() => number | string) | number | string>(commandNameResolvable: T): AllMatches<void>;
}
