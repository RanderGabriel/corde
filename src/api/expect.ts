import { expectMatchWithNot, MatchesWithNot } from "./matcher";

/**
 * Receives wich command will be tested.
 *
 * @param commandName Command name.
 *
 * @description Do not inform the command prefix if
 * it's already informed in **configs**
 *
 * @returns The **Compare** object, where will handle
 * the type of response is expected.
 */
export function expect(commandName: string): MatchesWithNot {
  return expectMatchWithNot(commandName);
}
