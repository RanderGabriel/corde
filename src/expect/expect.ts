import { VoidLikeFunction } from "../types";
import { ExpectMatchesWithNot, MatchWithNot } from "./matcher";

/**
 * Invalid declaration of a test matcher.
 *
 * @since 3.0
 */
export function expect(): void;
/**
 * Receives which command will be tested.
 *
 * Do not inform the command prefix if
 * it's already informed in **configs**
 *
 * @param commandNameResolvable Command name. (Empty strings will resolve failed test)
 *
 * @returns An object with all possible tests to be done
 * in the bot.
 *
 * @since 1.0
 */
export function expect<T extends VoidLikeFunction | number | string>(
  commandNameResolvable: T,
): MatchWithNot;
export function expect<T extends VoidLikeFunction | number | string>(
  commandNameResolvable?: T,
): MatchWithNot | void {
  if (commandNameResolvable == null) {
    return;
  }
  return new ExpectMatchesWithNot(commandNameResolvable);
}
