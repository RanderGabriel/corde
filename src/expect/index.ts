import { RoleMatches, MessageMatches } from "./matcher";
import { Expect, MacherContructorArgs } from "./types";

function getMessageMatchers(): string[] {
  return getFunctions(MessageMatches);
}

function getRoleMatchers(): string[] {
  return getFunctions(RoleMatches);
}

function getFunctions<T>(type: new (...args: any[]) => T): string[] {
  return Object.getOwnPropertyNames(type.prototype).filter(
    (propName) => propName !== "constructor" && typeof type.prototype[propName] === "function",
  );
}

const expectation = {
  not: {},
  inChannel: {},
  inGuild: {},
};

function createTestFunction(classType: any, params: MacherContructorArgs, functionName: string) {
  return (...args: any[]) => (new classType(params) as any)[functionName](...args);
}

function createTestsFromMatches<T>(
  names: string[],
  params: MacherContructorArgs,
  classType: new (...args: any[]) => T,
): [string, (...args: any[]) => any][] {
  return names.map((name) => {
    return [name, createTestFunction(classType, params, name)];
  });
}

function set(from: [string, (...args: any[]) => any][], to: any) {
  from.forEach((test) => {
    to[test[0]] = test[1];
  });
}

const messageTestNames = getMessageMatchers();
const roleMatchers = getRoleMatchers();

const _expect: any = <T extends (() => number | string) | number | string>(commandName: T) => {
  const messageTests = createTestsFromMatches(
    messageTestNames,
    { commandName, isNot: false },
    MessageMatches,
  );

  const isNotMessageTests = createTestsFromMatches(
    messageTestNames,
    { commandName, isNot: true },
    MessageMatches,
  );

  const roleTests = createTestsFromMatches(
    roleMatchers,
    { commandName, isNot: false },
    RoleMatches,
  );

  const isNotRoleTests = createTestsFromMatches(
    roleMatchers,
    { commandName, isNot: true },
    RoleMatches,
  );

  set(messageTests, expectation);
  set(isNotMessageTests, expectation.not);

  set(roleTests, expectation);
  set(isNotRoleTests, expectation.not);

  expectation.inChannel = (channelId: string) => {
    const channelMatchers: any = {
      not: {},
    };

    const messageTests = createTestsFromMatches(
      messageTestNames,
      { commandName, channelId },
      MessageMatches,
    );

    const isNotmessageTests = createTestsFromMatches(
      messageTestNames,
      { commandName, channelId, isNot: true },
      MessageMatches,
    );

    set(messageTests, channelMatchers);
    set(isNotmessageTests, channelMatchers.not);

    return channelMatchers;
  };

  expectation.inGuild = (guildId: string) => {
    const guildMatchers: any = {
      not: {},
    };

    const roleTests = createTestsFromMatches(
      messageTestNames,
      { commandName, guildId },
      RoleMatches,
    );

    const isNotRoleTests = createTestsFromMatches(
      messageTestNames,
      { commandName, guildId },
      RoleMatches,
    );

    set(roleTests, guildMatchers);
    set(isNotRoleTests, guildMatchers.not);

    return guildMatchers;
  };

  return expectation;
};

const messageTests = createTestsFromMatches(
  messageTestNames,
  { commandName: "", isNot: false, isCascade: true },
  MessageMatches,
);

const isNotMessageTests = createTestsFromMatches(
  messageTestNames,
  { commandName: "", isNot: true, isCascade: true },
  MessageMatches,
);

const roleTests = createTestsFromMatches(
  roleMatchers,
  { commandName: "", isNot: false, isCascade: true },
  RoleMatches,
);

const isNotRoleTests = createTestsFromMatches(
  roleMatchers,
  { commandName: "", isNot: true, isCascade: true },
  RoleMatches,
);

_expect.not = {};

set(messageTests, _expect);
set(isNotMessageTests, _expect.not);

set(roleTests, _expect);
set(isNotRoleTests, _expect.not);

const expect = _expect as Expect;

export { expect };
