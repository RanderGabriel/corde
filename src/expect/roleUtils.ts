import { Role } from "discord.js";
import { RoleIdentifier } from "../types";
import { buildReportMessage, formatObject } from "../utils";

class RoleUtils {
  createExpectedMessageForRoleData(roleIdentifier: RoleIdentifier) {
    if (!roleIdentifier) {
      return null;
    }

    if (roleIdentifier.id && roleIdentifier.name) {
      return `role with id ${roleIdentifier.id} or name '${roleIdentifier.name}'`;
    }

    if (roleIdentifier?.id) {
      return `role with id ${roleIdentifier.id}`;
    }

    if (roleIdentifier?.name) {
      return `role with name '${roleIdentifier.name}'`;
    }

    return null;
  }

  getErrorForUndefinedRoleData(roleIdentifier: RoleIdentifier) {
    if (!roleIdentifier) {
      return buildReportMessage(
        "expected: data to identifier the role (id or name)\n",
        "received: null",
      );
    }

    return null;
  }

  validateRole(role: Role | undefined, roleIdentifier: RoleIdentifier): string | undefined {
    if (!role) {
      const message = roleUtils.createExpectedMessageForRoleData(roleIdentifier);

      if (message) {
        return buildReportMessage(`expected: ${message}\n`, "received: null");
      }

      return buildReportMessage(
        "expected: a id or a name to identify the role\n",
        `received: ${formatObject(roleIdentifier)}`,
      );
    }

    return undefined;
  }

  getRoleData(roleIdentifier: string | RoleIdentifier) {
    let data: RoleIdentifier;
    if (typeof roleIdentifier === "string") {
      data = { id: roleIdentifier };
    } else {
      data = roleIdentifier as RoleIdentifier;
    }
    return data;
  }
}

const roleUtils = new RoleUtils();
export { roleUtils };
