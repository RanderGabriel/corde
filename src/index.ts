/**
 * @file Automatically generated by barrelsby.
 */

 export * from "./cli";
export * from "./erros/index";
