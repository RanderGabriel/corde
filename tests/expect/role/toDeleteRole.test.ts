import { Client } from "discord.js";
import MockDiscord from "../../mocks/mockDiscord";
import { initCordeClientWithChannel } from "../../testHelper";
import { MockEvents } from "../../mocks/mockEvents";
import { TestReport } from "../../../src/types";
import { ToDeleteRole } from "../../../src/expect/matches";
import { roleUtils } from "../../../src/expect/roleUtils";
import { buildReportMessage } from "../../../src/utils";

let mockDiscord = new MockDiscord();
let mockEvents: MockEvents;

describe("testing toDeleteRole function", () => {
  afterEach(() => {
    mockDiscord = new MockDiscord();
  });

  it("should return a passed report for a deleted role(isNot false)", async () => {
    const corde = initializeCorde();
    mockEvents.mockOnceRoleDelete();

    const isNot = false;
    const toDeleteRole = new ToDeleteRole(corde, "test", isNot);
    const report = await toDeleteRole.action({ id: "123" });

    const model: TestReport = {
      pass: true,
    };

    expect(report).toMatchObject(model);
  });

  it("should return a failed report for a deleted role(isNot true)", async () => {
    const corde = initializeCorde();
    mockEvents.mockOnceRoleDelete();

    const isNot = true;
    const toDeleteRole = new ToDeleteRole(corde, "test", isNot);
    const report = await toDeleteRole.action({ id: "123" });

    const model: TestReport = {
      pass: true,
      message: "",
    };
    expect(report).toMatchObject(model);
  });

  it("should return a failed report for a no deleted role (isNot false)", async () => {
    const corde = initializeCorde();
    mockEvents.mockOnceRoleDelete();
    corde.fetchRole = jest.fn().mockReturnValue(mockDiscord.role);

    const isNot = false;
    const roleData = { id: "123" };
    const toDeleteRole = new ToDeleteRole(corde, "test", isNot);
    const report = await toDeleteRole.action(roleData);

    const baseMessage = roleUtils.createExpectedMessageForMessageData(roleData);
    const message = buildReportMessage(`expected: ${baseMessage}`, `received: ${roleData}`);

    const model: TestReport = {
      pass: false,
      message,
    };
    expect(report).toMatchObject(model);
  });

  it("should return a passed report for a found role with property role.deleted = true role (isNot false)", async () => {
    const corde = initializeCorde();
    mockEvents.mockOnceRoleDelete();
    const role = mockDiscord.role;
    role.deleted = true;
    corde.fetchRole = jest.fn().mockReturnValue(role);

    const isNot = false;
    const roleData = { id: "123" };
    const toDeleteRole = new ToDeleteRole(corde, "test", isNot);
    const report = await toDeleteRole.action(roleData);

    const message = buildReportMessage(
      "expected: a role not deleted",
      `received: the role with id: ${role.id} and name ${role.name} is already deleted`,
    );

    const model: TestReport = {
      pass: false,
      message,
    };
    expect(report).toMatchObject(model);
  });

  it("should return a failed report for a found role with property role.deleted = true role (isNot true)", async () => {
    const corde = initializeCorde();
    mockEvents.mockOnceRoleDelete();
    const role = mockDiscord.role;
    role.deleted = true;
    corde.fetchRole = jest.fn().mockReturnValue(role);

    const isNot = true;
    const toDeleteRole = new ToDeleteRole(corde, "test", isNot);
    const report = await toDeleteRole.action({ id: "123" });

    const message = buildReportMessage(
      "expected: a role not deleted",
      `received: the role with id: ${role.id} and name ${role.name} is already deleted`,
    );

    const model: TestReport = {
      pass: false,
      message,
    };
    expect(report).toMatchObject(model);
  });

  it("should return a passed report for a deleted role(isNot true)", async () => {
    const corde = initializeCorde();
    mockEvents.mockOnceRoleDelete();
    corde.fetchRole = jest.fn().mockReturnValue(mockDiscord.role);

    const isNot = true;
    const toDeleteRole = new ToDeleteRole(corde, "test", isNot);
    const report = await toDeleteRole.action({ id: "123" });

    const model: TestReport = {
      pass: true,
    };
    expect(report).toMatchObject(model);
  });

  it("should fail due to inexistent role", async () => {
    const corde = initCordeClientWithChannel(mockDiscord, new Client());
    mockEvents = new MockEvents(corde, mockDiscord);
    mockEvents.mockOnceRoleDelete();

    corde.findRole = jest.fn().mockReturnValue(null);

    const isNot = true;
    const roleData = { id: "123" };
    const toDeleteRole = new ToDeleteRole(corde, "test", isNot);
    const report = await toDeleteRole.action(roleData);

    const baseMessage = roleUtils.createExpectedMessageForMessageData(roleData);
    const message = buildReportMessage(`expected: ${baseMessage}`, `received: ${roleData}`);

    const model: TestReport = {
      pass: false,
    };
    expect(report).toMatchObject(model);
  });
});

function initializeCorde() {
  const corde = initCordeClientWithChannel(mockDiscord, mockDiscord.client);

  corde.findRole = jest.fn().mockReturnValue(mockDiscord.role);
  corde.sendTextMessage = jest.fn().mockImplementation(() => {});
  corde.fetchRole = jest.fn().mockReturnValue(null);
  mockEvents = new MockEvents(corde, mockDiscord);
  return corde;
}
