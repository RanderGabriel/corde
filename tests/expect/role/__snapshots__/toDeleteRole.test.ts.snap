// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing toDeleteRole function should fail due to inexistent role 1`] = `
Object {
  "message": "    
    expected: role with id 123
    received: null
",
  "pass": false,
}
`;

exports[`testing toDeleteRole function should fail due to roleData null 1`] = `
Object {
  "message": "    
    expected: data to identifier the role (id or name)
    received: null
",
  "pass": false,
}
`;

exports[`testing toDeleteRole function should return a failed report for a deleted role(isNot true) 1`] = `
Object {
  "message": "    
    expected: role 819382540350652502 to not be deleted
",
  "pass": false,
}
`;

exports[`testing toDeleteRole function should return a failed report for a found role with property role.deleted = true role (isNot true) 1`] = `
Object {
  "message": "    
    expected: role 819382540350652502 not deleted
    received: role was deleted before call the command '!removeRole 123'
",
  "pass": false,
}
`;

exports[`testing toDeleteRole function should return a failed report for a no deleted role (isNot false) 1`] = `
Object {
  "message": "    
    expected: role 819382540350652502 to be deleted
",
  "pass": false,
}
`;

exports[`testing toDeleteRole function should return a passed report for a deleted role(isNot false) 1`] = `
Object {
  "pass": true,
}
`;

exports[`testing toDeleteRole function should return a passed report for a found role with property role.deleted = true role (isNot false) 1`] = `
Object {
  "message": "    
    expected: role 819382540350652502 not deleted
    received: role was deleted before call the command '!deleteRole 123'
",
  "pass": false,
}
`;
